### ISSUE TITLE
Semantics of toplevel let-bindings in classes

### ISSUE THREAD

**Issue by lthls** [ocaml/ocaml#13178]
I have a question that can be explained with this example (and some variations later):
```ocaml
module M1 = struct
  let () = print_endline "Before class"
  class c =
    let () = print_endline "Class init" in
    object end
  let () = print_endline "After class"
  let o1 = new c
  let o2 = new c
end
```
The manual currently specifies that the `"Class init"` line should be evaluated between the two other print statements, and nothing printed when evaluating `o1` and `o2` (see https://ocaml.org/manual/5.2/classes.html#sss:class-localdefs). The implementation mostly follows that.

However, the implementation is quite complex and in particular, adding a local open can trick it into believing that the let bindings are not toplevel anymore:
```ocaml
module M2 = struct
  let () = print_endline "Before class"
  class c =
    let open Unit in
    let () = print_endline "Class init" in
    object end
  let () = print_endline "After class"
  let o1 = new c (* Prints here *)
  let o2 = new c (* Prints here again *)
end
```
I will propose a PR that fixes this issue, but I'd like to know why we even bother with that corner case: making everything under the class definition evaluated at object creation time seems reasonable to me and easier to implement and specify.

In particular, in the following definitions `M4` and `M5` behave like `M2` (printing pushed at object creation time) despite being mostly equivalent to `M3` and `M1` respectively (which both print at class creation time):
```ocaml
module M3 = struct
  class with_param p = object end
  let () = print_endline "Before class"
  class c =
    let () = print_endline "Class init" in
    with_param ()
  let () = print_endline "After class"
  let o1 = new c
  let o2 = new c
end

module M4 = struct
  class with_param p = object end
  let () = print_endline "Before class"
  class c =
    (let () = print_endline "Class init" in
     with_param)
      ()
  let () = print_endline "After class"
  let o1 = new c
  let o2 = new c
end

module M5 = struct
  let () = print_endline "Before class"
  class c =
    (let () = print_endline "Class init" in
     object end : object end)
  let () = print_endline "After class"
  let o1 = new c
  let o2 = new c
end
```

Finally, the part in the manual about class application (https://ocaml.org/manual/5.2/classes.html#sss:class-app) is either wrong or misleading. The arguments passed to the class constructor are always evaluated at object creation time, not immediately, as can be seen on the following examples (in both cases, `"Class param"` is printed when creating the objects):
```ocaml
module M3_app = struct
  class with_param p = object end
  let () = print_endline "Before class"
  class c =
    let () = print_endline "Class init" in
    with_param (print_endline "Class param")
  let () = print_endline "After class"
  let o1 = new c
  let o2 = new c
end

module M4_app = struct
  class with_param p = object end
  let () = print_endline "Before class"
  class c =
    (let () = print_endline "Class init" in
     with_param)
      (print_endline "Class param")
  let () = print_endline "After class"
  let o1 = new c
  let o2 = new c
end
```

Cc @garrigue, our main reference on objects and classes.

**Comment by alainfrisch** [comment]
> I'd like to know why we even bother with that corner case: making everything under the class definition evaluated at object creation time seems reasonable to me and easier to implement and specify. 

I don't know the original motivation, but I can imagine that one reason is to allow variables scoped on the class, as in:

````ocaml
class c =
  let r = ref 0 in
  object
    val x = (incr r; !r)
    method x = x
  end
````

If the variable is lifted before the class definition, it's cumbersome to hide it for the rest of the module (I don't see how to do it without writing down the signature for the class).

**Comment by alainfrisch** [comment]
Somehow related: https://github.com/ocaml/ocaml/issues/4807

**Comment by lthls** [comment]
> If the variable is lifted before the class definition, it's cumbersome to hide it for the rest of the module (I don't see how to do it without writing down the signature for the class).

You can do it using modules, although it doesn't completely hide the reference (it's available through the module):
```ocaml
module Class_implem = struct
  let r = ref 0
  class c =
    object
      val x = (incr r; !r)
      method x = x
    end
end
class c = Class_implem.c
```
You can also shadow it:
```ocaml
let r = ref 0
class c =
  object
    val x = (incr r; !r)
    method x = x
  end
let r _ = assert false
```
Or use names that are obviously not meant to be reused:
```ocaml
let class_local_r = ref 0
class c =
  let r = class_local_r in
  object
    val x = (incr r; !r)
    method x = x
  end
```
But I think my preference would be to allow lifting, but only for explicitly annotated bindings:
```ocaml
class c =
  let[@class_init] r = ref 0 in
  object
    val x = (incr r; !r)
    method x = x
  end
(* Or with explicit syntax: *)
class c =
  let class r = ref 0 in
  object
    val x = (incr r; !r)
    method x = x
  end
```
The type-checker would then error out if it turns out the binding isn't in a liftable position (not at toplevel, or depends on previous, non-lifted variables).

**Comment by lthls** [comment]
> If the variable is lifted before the class definition, it's cumbersome to hide it for the rest of the module (I don't see how to do it without writing down the signature for the class).

And as a bonus (I don't think it passes the "not cumbersome" test, but it does completely hide the reference without writing the class type):
```ocaml
module F () = struct
  let r = ref 0
  class c_inner =
    object
      val x = (incr r; !r)
      method x = x
    end
end
include struct open F() class c = c_inner end
```
(It doesn't work very well if you need mutually recursive classes each with their own local variables though)

**Comment by gasche** [comment]
The use of eagerly-evaluated let-bindings in class definitions is a very intentional design choice that was certainly motivated by practical applications. Changing the semantics at this point sounds fairly dangerous, as we may break code in the wild without noticing. I would expect any sort of change proposal here to come with an analysis of the impact on existing OCaml codebases, for example in the opam-repository.

**Comment by lthls** [comment]
The opam repository does not contain many tests, so it's likely going to miss a lot of the potentially impacted users. But I guess we might start with that.
If you happen to have a script to run tests on many opam packages, I'm interested. Even without runtime tests, I could log whether it actually triggers or not and manually check if it would make a difference in practice.

**Comment by gasche** [comment]
It may be that grepping for `class` and a manual review would let you isolate candidates for interesting eager-let behavior without building or running any code. (Classes with parameters are ruled out, and classes of the form `= object` are ruled out; so you are looking for `class <name> = not(object)`.)

**Comment by gasche** [comment]
The following search on Github already gives some interesting results:
   `/^class [a-zA-Z_]* =$/ /^ *let/ lang:OCaml`
[Search results](https://github.com/search?q=%2F^class+[a-zA-Z_]*+%3D%24%2F+%2F^+*let%2F+lang%3AOCaml+&type=code).

**Comment by lthls** [comment]
There are indeed a number of pieces of code where the change of semantics may be observable.
A number of these results only bind functions, so the only observable change will be when the closure is allocated (if it is dynamically allocated).
I've noted 5 of them that are likely to be more fundamentally impacted:
- https://github.com/janestreet/lwt-async/blob/c738e6202c1c7409e079e513c7bdf469f7f9984c/src/text/lwt_read_line.ml#L35 looks like it's expecting the non-lifted semantics, though it's hard to tell given that the repo only contains a single top-level instanciation of the class. The class is exposed in the mli, with a comment that I interpret as expecting two different instantiations *not* to share their states.
- https://github.com/appliedfm/vstyle-tools/blob/4dc4c719ab604ba255e3b76c8adc2c57574211e8/src/doctree.ml#L66 is impacted, but I couldn't tell which semantics was intended or whether the change would be noticeable
- https://github.com/Kappa-Dev/KappaTools/blob/198998406788b99ae0d13caff64fa689db712409/webapp/route_root.ml#L26 looks like it probably expects the non-lifted semantics (the processes created at the start of the class are terminated by each instance, so I suspect they should be matched; on the other hand some of the processes are never terminated at all so maybe these ones are meant to be unique)
- https://github.com/s3team/uroboros/blob/c074cca980eb94cc16735f1e568b08bb308001fc/src/parser.ml#L15 has a shared reference. I expect it doesn't matter (the reference is reset when creating an instance, and I suspect that two instances never exist in parallel).
- https://github.com/0install/0install/blob/225587eef889a3082e0cc53fa64500f56cca0028/src/zeroinstall/console.ml#L18 looks like it expects a single instance of the class, so it's hard to say whether the bindings are expected to be run immediately or on object creation.

My conclusion is that it's not obvious that changing the semantics is harmless, but there is some evidence that the current semantics is not always the one users expect.

**Comment by gasche** [comment]
cc @garrigue who may have an opinion

**Comment by gasche** [comment]
In my opinion this is now part of the language specification, so we would need strong reasons to change it.

I consider the current `let open` behavior to be a bug, to be fixed.

**Comment by garrigue** [comment]
I agree with @gasche. For let bindings, this has been the specification from the beginning, so better not to change it.
For class arguments, the delayed evaluation is more surprising. It would be good to check the behavior in ocaml 2.04. This looks like a bug. I don't think anybody relies on that, though.
